c2rust-analyze/src/util.rs

```diff
@@ -353,3 +354,64 @@ pub fn is_null_const(constant: Constant) -> bool {
 
 pub trait PhantomLifetime<'a> {}
 impl<'a, T: ?Sized> PhantomLifetime<'a> for T {}
+
+/// Determine if `from` can be safely transmuted to `to`,
+/// which is defined as `*(from as *const To)` being a safe operation,
+/// where `from: *const From` and assuming `*from` already was safe.
+///
+/// Note that this is one-way, and is slightly different from [`core::mem::transmute`],
+/// and more similar to [`core::mem::transmute_copy`].
+///
+/// This forms a reflexive, transitive, and non-symmetric (one-way) relation, named `~` below.
+/// Formally, `A ~ B` iff `*a` and `*(a as *const B)` are safe, where `a: *const A`.
+///
+/// However, safe transmutability is difficult to check completely,
+/// so this function only checks a subset of it,
+/// with these formal rules for all types `A`, `B`:
+///
+/// * `A = B => A ~ B`
+/// * `A ~ B => *A ~ *B`
+/// * `uN ~ iN`, `iN ~ uN`, where `N` is an integer width
+/// * `[A] ~ A`
+/// * `[A; N] ~ A`, where `const N: usize`
+///
+/// Thus, [`true`] means it is definitely transmutable,
+/// while [`false`] means it may not be transmutable.
+pub fn is_transmutable_to<'tcx>(from: Ty<'tcx>, to: Ty<'tcx>) -> bool {
+    let transmutable_ints = || {
+        use IntTy::*;
+        use UintTy::*;
+        match (from.kind(), to.kind()) {
+            (ty::Uint(u), ty::Int(i)) | (ty::Int(i), ty::Uint(u)) => {
+                matches!(
+                    (u, i),
+                    (Usize, Isize) | (U8, I8) | (U16, I16) | (U32, I32) | (U64, I64)
+                )
+            }
+            _ => false,
+        }
+    };
+
+    let one_way_transmutable = || match from.kind() {
+        &ty::Array(from, _) | &ty::Slice(from) => is_transmutable_to(from, to),
```

Comment:
This appears to implement `A ~ B => [A] ~ B` rather than `[A] ~ A`.

Reply:
I meant to have the rule be `A ~ B => [A] ~ B` (and the same for arrays).  Would that change your analysis above in https://github.com/immunant/c2rust/pull/839#discussion_r1218720249 at all?

Reply:
Update the rules in 1317e034b6c8fe847ff693e8f8f44ce74a4bc83b.

Reply:
I thought we previously discussed the rule `A ~ B => [A] -> B`, concluded that it was unsound, and replaced it with the more limited `[A] -> A` instead.

Reply:
Oh, maybe.  Let me check.

Reply:
I'm pretty sure what we discussed before and that I removed were the slice to array and array to slice casts/transmutes.  I don't remember any issues from `A ~ B, N > 0 => [A; N] ~ B` (the slice version I've removed now).

Reply:
I thought about this more and realized you can derive `[A; N] ~ B` from `[A; N] ~ A` and `A ~ B`, so there's no new issue with the `[A; N] ~ B` version of the rule.